graphqlapi/src/main/java/io/stargate/graphql/persistence/schemafirst/SchemaSourceDao.java

```diff
@@ -0,0 +1,319 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.graphql.persistence.schemafirst;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import io.stargate.db.datastore.DataStore;
+import io.stargate.db.datastore.ResultSet;
+import io.stargate.db.datastore.Row;
+import io.stargate.db.query.BoundQuery;
+import io.stargate.db.query.Predicate;
+import io.stargate.db.query.builder.BuiltCondition;
+import io.stargate.db.query.builder.Replication;
+import io.stargate.db.schema.Column;
+import io.stargate.db.schema.ImmutableColumn;
+import io.stargate.db.schema.ImmutableTable;
+import io.stargate.db.schema.Keyspace;
+import io.stargate.db.schema.Table;
+import io.stargate.graphql.schema.schemafirst.migration.CassandraSchemaHelper;
+import io.stargate.graphql.schema.schemafirst.util.Uuids;
+import java.util.*;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SchemaSourceDao {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SchemaSourceDao.class);
+  public static final String KEYSPACE_NAME = "stargate_graphql";
+  public static final String TABLE_NAME = "schema_source";
+  @VisibleForTesting static final String NAMESPACE_COLUMN_NAME = "namespace";
+  @VisibleForTesting static final String VERSION_COLUMN_NAME = "version";
+  @VisibleForTesting static final String LATEST_VERSION_COLUMN_NAME = "latest_version";
+  @VisibleForTesting static final String CONTENTS_COLUMN_NAME = "contents";
+  @VisibleForTesting static final String APPLIED_COLUMN_NAME = "[applied]";
+
+  @VisibleForTesting
+  static final String DEPLOYMENT_IN_PROGRESS_COLUMN_NAME = "deployment_in_progress";
+
+  private static final int NUMBER_OF_RETAINED_SCHEMA_VERSIONS = 10;
+
+  @VisibleForTesting
+  static final Table EXPECTED_TABLE =
+      ImmutableTable.builder()
+          .keyspace(KEYSPACE_NAME)
+          .name(TABLE_NAME)
+          .addColumns(
+              ImmutableColumn.create(
+                  NAMESPACE_COLUMN_NAME, Column.Kind.PartitionKey, Column.Type.Varchar),
+              ImmutableColumn.create(
+                  VERSION_COLUMN_NAME,
+                  Column.Kind.Clustering,
+                  Column.Type.Timeuuid,
+                  Column.Order.DESC),
+              ImmutableColumn.create(
+                  CONTENTS_COLUMN_NAME, Column.Kind.Regular, Column.Type.Varchar),
+              ImmutableColumn.create(
+                  LATEST_VERSION_COLUMN_NAME, Column.Kind.Static, Column.Type.Timeuuid),
+              ImmutableColumn.create(
+                  DEPLOYMENT_IN_PROGRESS_COLUMN_NAME, Column.Kind.Static, Column.Type.Boolean))
+          .build();
+
+  private final DataStore dataStore;
+
+  public SchemaSourceDao(DataStore dataStore) {
```

Comment:
In theory I wholeheartedly agree. It was my plan to make everything async and I tried in [this PR](https://github.com/stargate/stargate/pull/544#issuecomment-758786043). Unfortunately I hit a wall because our internal API is not fully async, due to the fact that the Cassandra code we rely on was not designed that way either.

Reply:
Yea exactly, I think there should be a way to overcome this problem. Let's keep this commit for later. You can always wrap sync code into `CompletableFuture` (there should be no difference if the thread running it is HTTP one or the one from the executor, as you can have dedicated executors for this, that Cassandra code has no access to) or exec synchronously and then switch to async.